[RFC] Use DTS firmware partition to obsolete IMAGE_SIZE

adschm opened the pull request:
This is a first attempt to get rid of IMAGE_SIZE by reading firmware partition size from device tree and use it to check size (original proposal by @ynezz).

Currently I need to "apt install device-tree-compiler" on Debian 10 to have access to the `fdtget` function. However, this seems to be just a simple wrapper to access the kernel `fdget` function.

adschm committed:
ath79: add function to check image size based on firmware partition

So far, each device defines a variable IMAGE_SIZE, which is then
used to check whether the firmware images will match into the
firmware partition.
Since this variable has to be set manually, it frequently was just
copied from another device or set to the wrong value due to other
reasons.

To circumvent that and save us from fixing IMAGE_SIZE every couple
of years, this introduces a function to check image size directly
based on the firmware partition node from DTS.

This is based on an original idea by Petr Štetiar <ynezz@true.cz>

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ath79: update partition definition syntax for some devices

This updates partition syntax to follow the abundant scheme:

	alias: partition@offset {
		label = "label";
		...
	};

The alias is optional.
Despite ensuring consistency, this will allow to read the label from dtb
files.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ath79: remove IMAGE_SIZE variable

This removes the IMAGE_SIZE variable entirely from ath79 target
and replaces check-size with check-size-firmware whenever a
firmware partition is available.

To prevent copy-and-paste of IMAGE_SIZE in the future, this also
removes the variable for cases where NO firmware partition is
available. Since those are rare, the patch just adds the
numeric value to the check-size call then.

Despite check-size, the are also very rare cases of other commands
using it. As for check-size, we just use the plain value there...

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented on include/image-commands.mk:
```diff
@@ -265,6 +265,13 @@ define Build/check-size
 endef
 
+define Build/check-size-firmware
+	sh $(TOPDIR)/scripts/check-size-firmware.sh $(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb "$$(stat -c%s $@)" || { \
```
Here, I'm still thinking about what's the best shape for the script:
- One could change the script to just be a get-size-firmware.sh, so the check would move to the Makefile here. This would enable use of the script in other places, but would also make handling of errors less comfortable (missing firmware node would be treated as "image too big" ...)
- One could move the stat command to the script and just pass $@ as second parameter here. This would make it more complex to just call the script to learn about firmware partition sizes...

ynezz replied:
Indeed, it should be more generic, so it could be reused in other targets. Probably something like `get-dts-part-size.sh`.

> also make handling of errors less comfortable (missing firmware node would be treated as "image too big" ...)

We would use this script only if we know, that it's going to work with the DTS, right?

adrianschmutzler replied:
> We would use this script only if we know, that it's going to work with the DTS, right?

Yes, I'm just not sure whether everyone will think of that.
However, it will fail when there is no firmware node, so the user will be notified. He just won't see the correct message then.

ynezz commented on scripts/check-size-firmware.sh:
```diff
@@ -0,0 +1,22 @@
+DTB=$1
+IMAGE_FILE_SIZE=$2
+PARTITION_NODES="/ahb/spi/flash/partitions /spi/flash/partitions"
+FDGET=fdtget
```
This should be removed, and defined as `FDTGET:=fdtget` in `rules.mk`.

adrianschmutzler replied:
rules.mk is for Makefiles.
I will go backwards here and just hardcode "fdtget" it in the lines below.

adrianschmutzler replied:
Will the fdtget from your path be "in path" here, or will I have to use an explicit path?

ynezz commented on scripts/check-size-firmware.sh:
```diff
@@ -0,0 +1,22 @@
+
+DTB=$1
+IMAGE_FILE_SIZE=$2
+PARTITION_NODES="/ahb/spi/flash/partitions /spi/flash/partitions"
```
This should be generic, so passed as argument.

adrianschmutzler replied:
Done.

ynezz commented on scripts/check-size-firmware.sh:
```diff
@@ -0,0 +1,22 @@
+for part_node in $PARTITION_NODES; do
+	for partition in $($FDGET -l $DTB $part_node 2>/dev/null); do
+		name=$($FDGET $DTB $part_node/$partition label 2>/dev/null)
+		[ "$name" = "firmware" ] && {
```
I'm not sure if hardcoding `firmware` makes sense here, maybe it should be parameter as well?

adrianschmutzler replied:
Done.

ynezz commented:
Here [work-in-progress patch](https://github.com/ynezz/openwrt/commit/f4866fc3bc69c3c2ec7511841aed91444f1cddb8) which adds `tools/dtc` package which in turn provides host `fdtget` and `fdtdump` utils.

adschm committed:
ath79: merge partition definitions

In two cases, spi nodes and partition definitions were split
between DTSI and DTSes. Since this duplication is not heavy, but
complicates reading the files, this patch merges partition nodes
in the DTS files for those cases.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
Pushed an updated version:
- Changed script to only return the partition size and added more parameters to it
- Moved check-size-firmware to ath79: Since paths and partitions might be target specific, it is more reasonable to define check-size-* for each target.
- Split partition tidy-up on ath79 into two patches

Tested this with three cases:
1. If everything is fine, everything is fine
2. If firmware partition is too small, the usual "WARNING: ... is too small" is printed.
3. If no matching partition is found, an additional warning "WARNING: No $FW_PART partition found" is printed. This will return size=0, and since the logic in the Makefile cannot deal with that, an additional "too small" warning will be printed. Since this only happens when the device is misconfigured in the first place, I consider this a suitable compromise.

CodeFetch commented:
@adrianschmutzler What about adding one blocksize to the size constraint so that it's always possible to initialize JFFS2?

CodeFetch commented:
Another thing I wonder about: Does the JFFS2 start directly (the next byte) after the OpenWrt image/can it happen that it shares a block with it? If yes, if there is a power loss while writing to that common block, wouldn't it possibly brick the router in a way that even failsafe mode is not usable anymore? If yes, wouldn't it make sense to always pad OpenWrt images to blocksize?

adrianschmutzler commented:
@CodeFetch Maybe have a look at this discussion (which led me to create this PR in the first place):
https://github.com/openwrt/openwrt/pull/2226

I'm not sure what's the best approach to deal with this whole topic at the moment.

CodeFetch commented:
@adrianschmutzler I think mkresin's solution is the only viable. Adding a padding to blocksize in pad-rootfs before padjffs2 is actually being called (as far as I understood him) and adding a check here whether there is still one blocksize free compared to the size of the firmware partition. Now that we got rid of tiny devices this looks like a step forward.

adrianschmutzler commented:
Since I don't have the capacity to continue this at the moment, I close the PR and will reopen when I have made new progress.

@ynezz If you are annoyed by the fdtget commit in your staging, I might take it into my development branch for this project.
